File: src/main/java/me/imoltres/bbu/listeners/RespawnListener.java
package me.imoltres.bbu.listeners;

import me.imoltres.bbu.BBU;
import me.imoltres.bbu.data.player.BBUPlayer;
import me.imoltres.bbu.data.team.BBUTeam;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerRespawnEvent;

public class RespawnListener implements Listener {
    @EventHandler
    public void onRespawn(PlayerRespawnEvent e) {
        Player bukkitPlayer = e.getPlayer();
        BBUPlayer player = BBU.getInstance().getPlayerController().getPlayer(bukkitPlayer.getUniqueId());
        if (player == null)
            return;
        if (player.getTeam() == null)
            return;
        BBUTeam team = player.getTeam();
        if (bukkitPlayer.getBedSpawnLocation() != null) {
            bukkitPlayer.teleport(bukkitPlayer.getBedSpawnLocation());
            return;
        }
        if (team.hasBeacon()) {
            bukkitPlayer.teleport(team.getBeacon().toWorldPosition("world").toBukkitLocation().add(0.5, 0.5, 0.5));
        } else {
            bukkitPlayer.teleport(team.getCage().getSpawnPosition().toBukkitLocation());
        }
    }
}

File: src/main/java/me/imoltres/bbu/listeners/RespawnListener.kt
package me.imoltres.bbu.listeners

import me.imoltres.bbu.BBU
import me.imoltres.bbu.data.player.BBUPlayer
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerRespawnEvent


// @sinender (taken from the closed PR)
// --------------------------------------------------------
class RespawnListener : Listener {
    @EventHandler
    fun onRespawn(e: PlayerRespawnEvent) {
        val bukkitPlayer: Player = e.player
        val player: BBUPlayer = BBU.getInstance().playerController.getPlayer(bukkitPlayer.uniqueId) ?: return
        val team = player.team
        if (team == null) return
        if (bukkitPlayer.respawnLocation != null) {
            // dont handle teleporting if the respawn location is already set
            // player set their respawn location to a bed or something, so we should respect that
            return
        }

        // if the team has a beacon, teleport them to it, otherwise teleport them to their cage
        // for some reason hasBeacon returns true in grace period... what idiot coded that (me)
        if (team.cage != null) {
            e.respawnLocation = team.cage!!.spawnPosition.toBukkitLocation()
        }
    }
}

// --------------------------------------------------------